core/drivers/tpm2/tpm2_ptp_fifo.c

```diff
@@ -0,0 +1,438 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2022, Linaro Limited
+ */
+
+#include <assert.h>
+#include <drivers/tpm2/tpm2_chip.h>
+#include <drivers/tpm2/tpm2_cmd.h>
+#include <drivers/tpm2/tpm2_ptp_fifo.h>
+#include <kernel/delay.h>
+#include <trace.h>
+
+static bool tpm2_fifo_check_ops(const struct tpm2_ptp_phy_ops *ops)
+{
+	if (!ops || !ops->rx32 || !ops->tx32 || !ops->rx8 || !ops->tx8)
+		return false;
+
+	return true;
+}
+
+static int tpm2_fifo_wait_valid(struct tpm2_chip *chip, uint32_t loc)
+{
+	const struct tpm2_ptp_phy_ops *ops = chip->phy_ops;
+	int rc = -1;
+	uint64_t timeout_ref = timeout_init_us(chip->timeout_a * 1000);
+	uint8_t access = 0;
+
+	do {
+		rc = ops->rx8(chip, TPM2_ACCESS(loc), 1, &access);
+		if (rc)
+			return TPM2_ERR_IO;
+
+		if (access & TPM2_ACCESS_VALID)
+			return TPM2_OK;
+
+		mdelay(TPM2_TIMEOUT_RETRY_MS);
+	} while (!timeout_elapsed(timeout_ref));
+
+	return TPM2_ERR_TIMEOUT;
+}
+
+static bool tpm2_fifo_check_locality(struct tpm2_chip *chip, uint32_t loc)
```

Comment:
It seems this function does more than the name suggests. A comment somewhere on what's going on would be welcome.